File: cpp/src/server/photos_handler/photo_representations.cpp
/*
 WebServe
 Copyright 2018 Peter Pearson.

 Licensed under the Apache License, Version 2.0 (the "License");
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ---------
*/

#include "photo_representations.h"

PhotoRepresentations::PhotoRepresentations()
{

}

void PhotoRepresentations::addRepresentation(const PhotoRep& photoRepr)
{
	m_aRepresentations.emplace_back(photoRepr);
}

const PhotoRepresentations::PhotoRep* PhotoRepresentations::getFirstRepresentationMatchingCriteriaMaxDimension(unsigned int maxVal, bool returnSmallestIfNotFound) const
{
	// for the moment, do a brute-force search..
	// TODO: in the future we could try and order the representations size-wise and do a binary search, but given we have to satisfy both dimensions
	//       that might not be trivial in terms of ordering...

	unsigned int smallestValFound = 6000;
	unsigned int smallestIndex = -1;

	unsigned int index = 0;
	for (const PhotoRep& repr : m_aRepresentations)
	{
		unsigned int maxDim = std::max(repr.getWidth(), repr.getHeight());
		if (maxDim <= maxVal)
		{
			return &repr;
		}

		if (maxDim < smallestValFound)
		{
			smallestValFound = maxDim;
			smallestIndex = index++;
		}
	}

	if (!returnSmallestIfNotFound)
	{
		// we didn't find anything matching that criteria...
		return nullptr;
	}

	// otherwise, return the smallest we previously found...
	if (smallestIndex != -1)
	{
		const PhotoRepresentations::PhotoRep* pSmallestBackup = &m_aRepresentations[smallestIndex];
		return pSmallestBackup;
	}

	return nullptr;
}

const PhotoRepresentations::PhotoRep* PhotoRepresentations::getFirstRepresentationMatchingCriteriaMinDimension(unsigned int minVal, bool returnLargestIfNotFound) const
{
	// for the moment, do a brute-force search..
	// TODO: in the future we could try and order the representations size-wise and do a binary search, but given we have to satisfy both dimensions
	//       that might not be trivial in terms of ordering...

	unsigned int largestValFound = 1;
	unsigned int largestIndex = -1;

	unsigned int index = 0;
	for (const PhotoRep& repr : m_aRepresentations)
	{
		unsigned int minDim = std::min(repr.getWidth(), repr.getHeight());
		if (minDim >= minVal)
		{
			return &repr;
		}

		if (minDim > largestValFound)
		{
			largestValFound = minDim;
			largestIndex = index++;
		}
	}

	if (!returnLargestIfNotFound)
	{
		// we didn't find anything matching that criteria...
		return nullptr;
	}

	// otherwise, return the largest we previously found...
	if (largestIndex != -1)
	{
		const PhotoRepresentations::PhotoRep* pLargestBackup = &m_aRepresentations[largestIndex];
		return pLargestBackup;
	}

	return nullptr;
}

const PhotoRepresentations::PhotoRep* PhotoRepresentations::getLargestRepresentationMatchingCriteriaMaxDimension(unsigned int maxVal) const
{
	unsigned int largestValFound = 1;
	unsigned int largestIndex = -1;

	unsigned int index = 0;
	for (const PhotoRep& repr : m_aRepresentations)
	{
		unsigned int maxDim = std::max(repr.getWidth(), repr.getHeight());

		if (maxDim <= maxVal && maxDim > largestValFound)
		{
			largestValFound = maxDim;
			largestIndex = index++;
		}
		else
		{
			index++;
		}
	}

	// return the largest we found...
	if (largestIndex != -1)
	{
		const PhotoRepresentations::PhotoRep* pLargestBackup = &m_aRepresentations[largestIndex];
		return pLargestBackup;
	}

	return nullptr;
}

const PhotoRepresentations::PhotoRep* PhotoRepresentations::getSmallestRepresentationMatchingCriteriaMinDimension(unsigned int minVal) const
{
	unsigned int smallestValFound = 12000;
	unsigned int smallestIndex = -1;

	unsigned int index = 0;
	for (const PhotoRep& repr : m_aRepresentations)
	{
		unsigned int maxDim = std::max(repr.getWidth(), repr.getHeight());

		if (maxDim >= minVal && maxDim < smallestValFound)
		{
			smallestValFound = maxDim;
			smallestIndex = index++;
		}
		else
		{
			index++;
		}
	}

	// return the smallest we found...
	if (smallestIndex != -1)
	{
		const PhotoRepresentations::PhotoRep* pSmallestBackup = &m_aRepresentations[smallestIndex];
		return pSmallestBackup;
	}

	return nullptr;
}

File: rs/src/photos_handler/photo_representations.rs
/*
 WebServe (Rust port)
 Copyright 2021 Peter Pearson.

 Licensed under the Apache License, Version 2.0 (the "License");
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ---------
*/


#[derive(Clone, Debug)]
pub struct PhotoRep {

pub relative_file_path:     String,
pub width:                  u16,
pub height:                 u16,
pub aspect_ratio:           f32,
}

impl PhotoRep {
    pub fn new(relative_path: &str, width: u16, height: u16) -> PhotoRep {
        let aspect_ratio = (width as f32) / (height as f32);
        PhotoRep { relative_file_path: relative_path.to_string(), width, height, aspect_ratio }
    }
}

#[derive(Clone, Debug)]
pub struct PhotoRepresentations {

    pub representations:    Vec<PhotoRep>
}

impl PhotoRepresentations {
    pub fn new() -> PhotoRepresentations {
        PhotoRepresentations { representations: Vec::with_capacity(0 )}
    }

    pub fn add_representation(&mut self, photo_rep: PhotoRep) {
        self.representations.push(photo_rep);
    }

    pub fn get_smallest_representation_matching_criteria_min_dimension(&self, min_val: u32) -> Option<&PhotoRep> {

        let mut smallest_val_found = 12000_u16;
        let mut smallest_index = -1i32;

        let mut index = 0;
        for repr in &self.representations {
            let max_dim = std::cmp::max(repr.width, repr.height);

            if max_dim >= min_val as u16 && (max_dim as u16) < smallest_val_found {
                smallest_val_found = max_dim;
                smallest_index = index;
            }
            index += 1;
        }

        if smallest_index == -1 {
            return None;
        }
        else {
            return Some(&self.representations[smallest_index as usize]);
        }
    }

    pub fn get_first_representation_matching_criteria_min_dimension(&self, min_val: u32, return_largest_if_not_found: bool) -> Option<&PhotoRep> {
        let mut largest_val_found = 1_i32;
        let mut largest_index = -1_i32;

        let mut index = 0;
        for repr in &self.representations {
            let min_dim = std::cmp::min(repr.width, repr.height);
            if min_dim >= min_val as u16 {
                return Some(repr);
            }

            if min_dim > largest_val_found as u16 {
                largest_val_found = min_dim as i32;
                largest_index = index;
            }
            index += 1;
        }

        if !return_largest_if_not_found {
            // we didn't find anything...
            return None;
        }

        // otherwise, return the largest we previously found...
        if largest_index != -1 {
            return Some(&self.representations[largest_index as usize]);
        }

        return None;
    }
}
